base/sparse/higherorderfns.jl

```diff
@@ -969,99 +963,70 @@ broadcast(f::Tf, A::SparseMatrixCSC, ::Type{T}) where {Tf,T} = broadcast(x -> f(
-_containertype(::Type{<:StructuredMatrix}) = PromoteToSparse
-broadcast_indices(::Type{PromoteToSparse}, A) = indices(A)
+Broadcast.rule(::Type{<:StructuredMatrix}) = PromoteToSparse
 
-# combinations explicitly involving Tuples and PromoteToSparse collections
-# divert to the generic AbstractArray broadcast code
-promote_containertype(::Type{PromoteToSparse}, ::Type{Tuple}) = Array
-promote_containertype(::Type{Tuple}, ::Type{PromoteToSparse}) = Array
-# combinations involving scalars and PromoteToSparse collections continue in the promote-to-sparse funnel
-promote_containertype(::Type{PromoteToSparse}, ::Type{Any}) = PromoteToSparse
-promote_containertype(::Type{Any}, ::Type{PromoteToSparse}) = PromoteToSparse
-# combinations involving sparse arrays and PromoteToSparse collections continue in the promote-to-sparse funnel
-promote_containertype(::Type{PromoteToSparse}, ::Type{AbstractSparseArray}) = PromoteToSparse
-promote_containertype(::Type{AbstractSparseArray}, ::Type{PromoteToSparse}) = PromoteToSparse
-# combinations involving Arrays and PromoteToSparse collections continue in the promote-to-sparse funnel
-promote_containertype(::Type{PromoteToSparse}, ::Type{Array}) = PromoteToSparse
-promote_containertype(::Type{Array}, ::Type{PromoteToSparse}) = PromoteToSparse
-# combinations involving Arrays and sparse arrays continue in the promote-to-sparse funnel
-promote_containertype(::Type{AbstractSparseArray}, ::Type{Array}) = PromoteToSparse
-promote_containertype(::Type{Array}, ::Type{AbstractSparseArray}) = PromoteToSparse
+Broadcast.rule(::Type{SPVM}, ::Type{Broadcast.Bottom0d})                = PromoteToSparse
+Broadcast.rule(::Type{SPVM}, ::Type{Broadcast.BottomVector})            = PromoteToSparse
+Broadcast.rule(::Type{SPVM}, ::Type{Broadcast.BottomMatrix})            = PromoteToSparse
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{Broadcast.Scalar})       = PromoteToSparse
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{Broadcast.Bottom0d})     = PromoteToSparse
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{Broadcast.BottomVector}) = PromoteToSparse
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{Broadcast.BottomMatrix}) = PromoteToSparse
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{SPVM})                   = PromoteToSparse
+# Combinations of sparse arrays and higher-dimensional arrays default to the generic infrastructure.
+# In this case it's best to keep the same argument order as above, so that we don't get ambiguities
+# or conflicting rules
+Broadcast.rule(::Type{SPVM}, ::Type{Broadcast.BottomArray{N}}) where N            = Broadcast.BottomArray{N}
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{Broadcast.BottomArray{N}}) where N = Broadcast.BottomArray{N}
+# Tuples lead to dense outputs
+Broadcast.rule(::Type{SPVM}, ::Type{Tuple})                             = Broadcast.BottomArray{2}
+Broadcast.rule(::Type{PromoteToSparse}, ::Type{Tuple})                  = Broadcast.BottomArray{2}
```

Comment:
IIUC, the result could be one- or two-dimensional?

Reply:
You're right, thanks. I will check whether my preservation of dimensionality in `BottomArray` basically forces people to keep track of it (I'm not sure we want that), or whether people can supply just `BottomArray`.

Reply:
Avoiding requiring extenders to field that additional complexity would be lovely. (That desire partially motivated the unhappy indirection layering in the existing code.)